L1Trigger/TrackTrigger/plugins/TTStubBuilder.cc

```diff
@@ -5,12 +5,86 @@
+ * \author Ian Tomalin
+ * \date 2013 - 2020
  *
  */
 
 #include "L1Trigger/TrackTrigger/plugins/TTStubBuilder.h"
 
+template <>
+void TTStubBuilder<Ref_Phase2TrackerDigi_>::updateStubs(
+    const edm::OrphanHandle<edmNew::DetSetVector<TTCluster<Ref_Phase2TrackerDigi_>>>& clusterHandle,
+    const edmNew::DetSetVector<TTStub<Ref_Phase2TrackerDigi_>>& inputEDstubs,
+    edmNew::DetSetVector<TTStub<Ref_Phase2TrackerDigi_>>& outputEDstubs) const {
+  /// Now, correctly reset the output
+  typename edmNew::DetSetVector<TTStub<Ref_Phase2TrackerDigi_>>::const_iterator stubDetIter;
+
+  for (stubDetIter = inputEDstubs.begin(); stubDetIter != inputEDstubs.end(); ++stubDetIter) {
+    /// Get the DetId and prepare the FastFiller
+    DetId thisStackedDetId = stubDetIter->id();
+    typename edmNew::DetSetVector<TTStub<Ref_Phase2TrackerDigi_>>::FastFiller outputFiller(outputEDstubs,
+                                                                                           thisStackedDetId);
+
+    /// detid of the two components.
+    ///This should be done via a TrackerTopology method that is not yet available.
+    DetId lowerDetid = thisStackedDetId + 1;
+    DetId upperDetid = thisStackedDetId + 2;
+
+    /// Get the DetSets of the clusters
+    edmNew::DetSet<TTCluster<Ref_Phase2TrackerDigi_>> lowerClusters = (*clusterHandle)[lowerDetid];
+    edmNew::DetSet<TTCluster<Ref_Phase2TrackerDigi_>> upperClusters = (*clusterHandle)[upperDetid];
+
+    /// Get the DetSet of the stubs
+    edmNew::DetSet<TTStub<Ref_Phase2TrackerDigi_>> theseStubs = inputEDstubs[thisStackedDetId];
+
+    /// Prepare the new DetSet to replace the current one
+    /// Loop over the stubs
+    typename edmNew::DetSet<TTCluster<Ref_Phase2TrackerDigi_>>::iterator clusterIter;
+    typename edmNew::DetSet<TTStub<Ref_Phase2TrackerDigi_>>::iterator stubIter;
+    for (stubIter = theseStubs.begin(); stubIter != theseStubs.end(); ++stubIter) {
+      /// Create a temporary stub
+      TTStub<Ref_Phase2TrackerDigi_> tempTTStub(stubIter->getDetId());
+
+      /// Compare the clusters stored in the stub with the ones of this module
+      const edm::Ref<edmNew::DetSetVector<TTCluster<Ref_Phase2TrackerDigi_>>, TTCluster<Ref_Phase2TrackerDigi_>>&
+          lowerClusterToBeReplaced = stubIter->clusterRef(0);
+      const edm::Ref<edmNew::DetSetVector<TTCluster<Ref_Phase2TrackerDigi_>>, TTCluster<Ref_Phase2TrackerDigi_>>&
+          upperClusterToBeReplaced = stubIter->clusterRef(1);
+
+      bool lowerOK = false;
+      bool upperOK = false;
+
+      for (clusterIter = lowerClusters.begin(); clusterIter != lowerClusters.end() && !lowerOK; ++clusterIter) {
+        if (clusterIter->getHits() == lowerClusterToBeReplaced->getHits()) {
+          tempTTStub.addClusterRef(edmNew::makeRefTo(clusterHandle, clusterIter));
+          lowerOK = true;
+        }
+      }
+
+      for (clusterIter = upperClusters.begin(); clusterIter != upperClusters.end() && !upperOK; ++clusterIter) {
```

Comment:
Identical response as previous comment to https://github.com/cms-sw/cmssw/pull/30648#discussion_r454945562

Reply:
@tomalin , I just have signed the PR, however, thinking more I would say, that this loop is potentially dangerous, because "upperClusters.end() && !upperOK" may never happens. If I am wrong and this happens in all cases, why you check "!upperOK" at each iteration? If it is needed to interrupt the loop earlier it is possible to do "break" inside the loop.